### ISSUE TITLE
set a timeout on a request?

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/waterline#933]
is there a way to set a timeout on a database call? 

trying to figure out if this is something Waterline should provide, or whether I should just add a setTimeout call with some kind of canceler if the query doesn't complete in time... (is there a way to cancel queries)?

**Comment by devinivy** [comment]
There's no way to set this, but it's a good idea!  I'm going to mark this as a feature request for the community to remark on.  It's possible that this job should be given to the adapters, but waterline could theoretically handle it in some ways too.

**Comment by tjwebb** [comment]
> is there a way to set a timeout on a database call?

This is supported by some adapters, e.g. https://github.com/cnect/sails-sqlserver (via `requestTimeout`), and other adapters whose drivers accept that kind of config value. But no, there's currently no standard waterline way of doing this.

+1 good idea

**Comment by dmarcelino** [comment]
This makes sense: :+1: .

I'm thinking it would be good to set this up via config, where it would be possible to have different configs for different DBs.

**Comment by kevinburkeshyp** [comment]
As far as I could see there are two use cases:

- my API must respond within 30 seconds, so set the maximum length for a DB query to 25-26 seconds (the DB wide setting)

- some (or more) individual endpoints has a very strict latency requirement, or we don't care if one particular query takes a very long time to run (individual query setting)

Personally I care more about the individual query setting than the global setting.

**Comment by alin-roman-zitec** [comment]
If you don't care about leaving the query running, which is not recommended, you can timeout using bluebird's timeout method:

    var Promise = require('bluebird');
    
    User.find().where({ somePropertyNotIndexed: 'some value' })
    .timeout(30 * 1000)
    .catch(Promise.TimeoutError, function() {
        req.serverError('Query timed out');
    });

I don't know if you can abort a running query, but if you can, you want to abort it in the TimeoutError catch.

**Comment by mateeyow** [comment]
+1

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
